packages/core/src/browser/widgets/select-component.tsx

```diff
@@ -0,0 +1,301 @@
+            this.hide(true);
+        }
+        ev.stopPropagation();
+        ev.nativeEvent.stopImmediatePropagation();
+    }
+
+    protected handleClickEvent(event: React.MouseEvent<HTMLElement>): void {
+        this.toggleVisibility();
+        event.stopPropagation();
+        event.nativeEvent.stopImmediatePropagation();
+    }
+
+    protected toggleVisibility(): void {
+        if (!this.fieldRef.current) {
+            return;
+        }
+        if (!this.state.dimensions) {
+            const rect = this.fieldRef.current.getBoundingClientRect();
+            this.setState({
+                dimensions: {
+                    top: rect.top + rect.height,
+                    left: rect.left,
+                    width: rect.width,
+                    parentHeight: rect.height
+                },
+            });
+        } else {
+            this.hide(true);
+        }
+    }
+
+    protected hide(reset?: boolean): void {
+        this.setState({
+            dimensions: undefined,
+            selected: reset ? this.state.original : this.state.selected,
+            original: reset ? this.state.original : this.state.selected
+        });
+    }
+
+    protected renderDropdown(): React.ReactNode {
+        if (!this.state.dimensions) {
+            return;
+        }
+        if (this.mountedListeners.size === 0) {
+            // Only attach our listeners once we render our dropdown menu
+            this.attachListeners();
+            // We can now also calculate the optimal width
+            this.optimalWidth = this.getOptimalWidth();
+            this.optimalHeight = this.getOptimalHeight(Math.max(this.state.dimensions.width, this.optimalWidth));
+        }
+        const clientHeight = document.getElementById('theia-app-shell')!.getBoundingClientRect().height;
+        const invert = this.optimalHeight > clientHeight - this.state.dimensions.top;
+        const { options } = this.props;
+        const { selected } = this.state;
+        const description = selected !== undefined && options[selected].description;
+        const markdown = selected !== undefined && options[selected].markdown;
+        const items = options.map((item, i) => this.renderOption(i, item));
+        if (description) {
+            let descriptionNode: React.ReactNode | undefined;
+            const className = 'theia-select-component-description';
+            if (markdown) {
+                descriptionNode = <div key="description" className={className}
+                    dangerouslySetInnerHTML={{ __html: DOMPurify.sanitize(description) }} />; // eslint-disable-line react/no-danger
+            } else {
+                descriptionNode = <div key="description" className={className}>
+                    {description}
+                </div>;
+            }
+            if (invert) {
+                items.unshift(descriptionNode);
+            } else {
+                items.push(descriptionNode);
+            }
```

Comment:
It would be slightly more efficient to wait till all items had been added and then `.reverse()` if necessary.

Reply:
It might be, but that would affect the order in which the selection options appear too. They should stay the same independent of whether we are in inverted mode or not. Only the description should change it's place.

Reply:
Oops, fair point - I thought this was applying to every option, not just the description.